### PR TITLE
Add WaitGroup synchronization primitive

### PR DESCRIPTION
**Pull request by ysbaddaden**
This is more efficient than creating a `Channel(Nil)` and looping to receive N messages: we don't need a queue, only a counter, and we can avoid spurious wake ups of the main fiber and resume it only once.

See the documentation for examples and more details.

### CODE REVIEW

**Review by straight-shoota — CHANGES_REQUESTED** [review]
I think we'll need a bunch more tests for this. E.g. for multiple `#wait` calls, fibers adding more fibers, `#add` with negative delta or `#done` called before `#wait`. The latter two could both result in `@counter < 0` which is a relevant invariant to verify.

The Go implementation has some test cases that we could take inspiration from: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/sync/waitgroup_test.go

**Comment by straight-shoota** [comment]
As food for thought: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/sync/waitgroup.go
Go seems to merge the lock and counter int a single atomic field. Which makes sense considering that a spin lock is really just another atomic counter. You probably won't need the full bit width for either of them. This removes the need to keep track of waiting fibers, they just stay blocked on the lock 🤯 
Maybe we could explore this in a future refactor?

**Comment by ysbaddaden** [comment]
Go merges the counter (i32) and the number of waiters (u32) into a single atomic (u64) which is a neat idea, then uses a semaphore to suspend the goroutines.

What I'm curious about is:
- how is the semaphore implemented?
- what about architectures without a 64-bit atomic?

File: src/wait_group.cr
Comment (ysbaddaden) [inline src/wait_group.cr:50]: FIXME: `n` may be negative :face_in_clouds:

**Comment by ysbaddaden** [comment]
I'm afraid following on Go won't be possible because their implementation takes advantage of the atomics returning the new value (e.g. add to counter => returns new counter + waiters), but Crystal relies on LLVM atomics that always return the old value, which... is often pointless :disappointed: 

For example, to support `add(-5)` I must do the math twice :facepalm: 

```
old_value = @counter.add(n)
new_value = @counter + n
```

File: src/wait_group.cr
Comment (yxhuvud) [inline src/wait_group.cr:24]: Somewhere, it would be nice to have some sort of text dealing with exceptions here. Or some description of who is expected to use this. Because if `do_something` raise an exception you get a stacktrace printed to stderr, but the program will neither finish nor crash. This is very much not a fantastic developer experience, especially not when run in an environment where the waitgroup itself is run in a fiber (like in a webserver).

Is this intended for end users to use directly or is it just a building block to build tools with that is less problematic? I know I will pick it up and use it for the latter, but some warning that it may create situations that are unfun to debug would be nice. This is a bigger problem than in go (which also have wait groups and expect end users to use them) because go don't have exceptions that will bobble up through the stack.

The good parts is that this code makes it a lot easier to build things on top of it that take care of this and doesn't have the sharp corners.

Comment (straight-shoota) [reply src/wait_group.cr:24]: Eventually, I would expect this to be mainly a building block for higher-level mechanics and users wouldn't deal with it directly unless implenting low-level concurrency stuff.

Comment (ysbaddaden) [reply src/wait_group.cr]: You're right, I changed the example to wall `wg.done` inside an `ensure` block. It's a safer default example :)

File: spec/std/wait_group_spec.cr
Comment (ysbaddaden) [inline spec/std/wait_group_spec.cr:45]: In Minitest I would call `flunk message` to fail a test. Is there an equivalent in spec?

Comment (HertzDevil) [reply spec/std/wait_group_spec.cr:45]: [`fail`](https://crystal-lang.org/api/master/Spec/Methods.html#fail%28msg%2Cfile%3D__FILE__%2Cline%3D__LINE__%29-instance-method)

**Review by beta-ziliani — APPROVED** [review]

File: src/wait_group.cr
Comment (alexkutsan) [inline src/wait_group.cr:83]: I like this donation very much! It will be very useful in many cases.

One proposal that probably can be done later on as a separate improvement - is to make the `wait` method compatible with `select` to support the following snippet: 

```crystal
select
  when wg.wait
    puts "All fibers done"
  when timeout(X.seconds)
    puts "Some fiber stuck"
end
```

File: spec/std/wait_group_spec.cr
Comment (straight-shoota) [inline spec/std/wait_group_spec.cr:49]: **thought:** I was wondering if the receive action really needs to run 16 times. Even a single receive would fail the spec.
So I *think* this could technically be unrolled to `15.times { wg2.done }; select ...; wg2.done`. Not sure if it improves much, but it might have a slight chance to describe the purpose a bit more clearly?

File: src/wait_group.cr
Comment (straight-shoota) [inline src/wait_group.cr:54]: **suggestion:** "in which case the behaviour is identical to `#done`" seems a bit odd, because the number of decrements is adjustable. So I'd suggest to just copy the description here.
```suggestion
  # A negative value decrements the counter. When the counter reaches zero,
  # all waiting fibers will be resumed.
  # Raises `RuntimeError` if the counter reaches a negative value.
```
If we want to avoid duplication, I'd suggest to shorten the docs for `#done` to ``Same as `add(-1)` ``.

File: src/wait_group.cr
Comment (straight-shoota) [inline src/wait_group.cr:68]: **suggestion:** Delegate the details of enqueing to `@waiting`. We have a special data type for this, so it's easy to add a dedicated method. This would separate different levels of implementation depth
```suggestion
      @waiting.enqueue_all
```

File: src/wait_group.cr
Comment (ysbaddaden) [reply src/wait_group.cr:83]: @bararchy We're missing a generic mechanism for timeouts... but we could abstract how it's implemented for `select` so that could be doable.

That doesn't mean we can't _also_ integrate with `select`: we could wait on channel(s) + waitgroup(s) + timeout. Now, I'm not sure how to do that.

File: spec/std/wait_group_spec.cr
Comment (ysbaddaden) [reply spec/std/wait_group_spec.cr:49]: To be honest I'm mostly following Go's test (which is quite smart). I read it as:

- I musn't receive a message (didn't even start to decrement);
- I decrement by one;
- I still musn't receive a message;
- ...
- I decrement by one (counter reaches zero)
- I must receive all messages.

I think this behavior sounds more realistic and yields fibers between each call to `#done` which may help trigger a potential bug, otherwise we'd just merely decrement a counter in a loop.

File: src/wait_group.cr
Comment (ysbaddaden) [reply src/wait_group.cr:68]: That would be nice, but `@waiting` is a generic `Crystal::PointerLinkedList(T)` data structure. The `T` are currently types that always wrap a `Fiber` but they have different interfaces and behavior: fifo queue with arbitrary deletion + full iteration.

I thus went for a generic method to iterate+clear the queue, following on the `#consume_each` method from `Crystal::ThreadLocalValue` that does just that. We might be able to reuse it in Channel when we close it (probably not necessary).

**Review by straight-shoota — APPROVED** [review]

File: src/wait_group.cr
Comment (ysbaddaden) [reply src/wait_group.cr:84]: I have a [commit to support WaitGroup in `select` expressions](https://github.com/crystal-lang/crystal/commit/dcfd2ad6bd7dd05b74a4db9cd0b6febbd035d339).

The integration wasn't complex after I understood how SelectAction and SelectContext are working, but the current implementation is very isolated to Channel (on purpose). Maybe the integration is not a good idea, but if proves to be a good idea, we might want to extract the `select` logic from `Channel` to the `Crystal` namespace.

I'll open a pull request after this one is merged, so we can have a proper discussion.

**Comment by ysbaddaden** [comment]
I enabled the waitgroup spec for the interpreter (and disabled the stress test because the interpreter is too slow)... and the interpreter crashed on CI:

```
FATAL: can't resume a running fiber: #<Fiber:0x7f7b0181baa0: main>
  from src/crystal/scheduler.cr:148:7 in 'swapcontext'
  from src/compiler/crystal/interpreter/interpreter.cr:354:7 in 'interpret'
  from src/indexable.cr:574:11 in '->'
  from src/fiber.cr:146:11 in 'run'
  from ???
```

**Comment by straight-shoota** [comment]
This failure consistently appears in https://github.com/crystal-lang/crystal/pull/14122 as well.

**Comment by straight-shoota** [comment]
There's a new RFC about the `WaitGroup` API: https://github.com/crystal-lang/rfcs/pull/3
It presents the content of this PR in a more structured manner.

**Comment by ysbaddaden** [comment]
Updated to follow the latest changes in [RFC #2](https://github.com/crystal-lang/crystal/pull/2): resume waiting fibers when the counter becomes negative + make `#wait` raise when the counter is negative.

**Comment by Sija** [comment]
`src/none.cr` seems to be added by accident?

File: src/wait_group.cr
Comment (Sija) [inline src/wait_group.cr:91]: Suggestion: might be a bit more readable.

```suggestion
    case @counter.get
    when .negative?
      raise RuntimeError.new("Negative WaitGroup counter")
    when .zero?
      return
    when .positive?
```

File: src/wait_group.cr
Comment (straight-shoota) [reply src/wait_group.cr:53]: Please refer to the RFC for a semantic discussion of this feature: https://github.com/crystal-lang/rfcs/pull/3

The TL;DR: We need `#add(Int32)` and the parameter can take a negative value which we must handle *somehow*. We could just raise, but decrementing is just as fine and valid within the data model.

I'm not aware of any example where you would want to decrement more than one, but there might be legitimate use cases where this makes sense.

File: src/wait_group.cr
Comment (ysbaddaden) [reply src/wait_group.cr:76]: That would duplicate the general WaitGroup example.

File: spec/std/wait_group_spec.cr
Comment (RX14) [inline spec/std/wait_group_spec.cr:43]: Doesn't this spec have a race condition where the waiter could be resumed after the first `wg.done` without raising, if these specs are run under a multithreaded scheduler?

File: src/wait_group.cr
Comment (RX14) [inline src/wait_group.cr:99]: I'm not convinced this shouldn't also check for a negative counter here.

Additionally, perhaps it's better (and simpler) to always acquire the lock given the overwhelmingly likely case is that the fiber will wait.

**Review by RX14 — CHANGES_REQUESTED** [review]

**Comment by RX14** [comment]
At the cost of using a `compare_and_set` loop, it would be possible to "saturate" counter decrements at 0 and provide a guarantee all future addition operations on a waitgroup that transitions to the `0` state will raise. This increases correctness, but will decrease throughput of a heavily contended waitgroup.

File: spec/std/wait_group_spec.cr
Comment (straight-shoota) [reply spec/std/wait_group_spec.cr:43]: There are probably a lot of specs that have race conditions like this. And it's hard to protect them properly because that would require using the primitives that are being tested.
So I think it might be best to ensure the spec runs in a single-threaded scheduler.

Comment (ysbaddaden) [reply spec/std/wait_group_spec.cr:43]: Damn, you're right, _and_ the test is wrong: the first `#done` is the one that resumes the waiting fiber. I guess I should manipulate the counter back to `0` then call `#done`.

**Comment by ysbaddaden** [comment]
Thanks @RX14. I'm noticing more edge cases. For example we could reach a negative number (raises) then a concurrent fiber would increment back to a positive number and fail to raise because the new counter is positive, also impacting the resumed waiting fibers that may continue despite the waitgroup being left in an invalid state.

I'll likely to add a CAS loop, just not saturating at zero, but keep the negative number to detect the invalid state.

File: src/wait_group.cr
Comment (straight-shoota) [inline src/wait_group.cr:63]: **nitpick:** It might be better to initialize to `nil` because then the compiler will complain about a logic error that could leave this variable being unassigned a real value.

**Comment by ysbaddaden** [comment]
I eased out corner edges:

- can't increment from a negative counter anymore (forever invalid state);
- raises on positive counter after resume (confused state).

I'm a bit torn about the last one: the situation can happen when the counter reached zero, enqueued waiters, and continued to increment, which is invalid, yet there is a race condition when reusing a WaitGroup with at least 2 waiters: fiber A reuses the WaitGroup (i.e. increments) before fiber B resumes (positive counter -> raise).

Oh, the race condition would also trigger with a negative counter (lower probability but could happen), so the problem is reusing the object before all waiters are properly resumed.

Ah, the joys of writing a synchronization primitive :feelsgood:

**Comment by RX14** [comment]
My highest concern is deadlocks: any condition where the counter remains on zero but fails to resume a waiter. There are race conditions when the waitgroup is used weirdly, but they do not cause a wrong count or a deadlock so can fail in a better way (raising).

File: src/wait_group.cr
Comment (ysbaddaden) [reply src/wait_group.cr]: I'm not fond of introducing nilables just for the sake of it (we immediately call `#not_nil!`).

I tried alternatives, like `val = loop { val = x; break val }` but "can't use variable inside assignment" (despite the value being defined in the block, so it could be understood as shadowing the variable). I didn't want to use a `tmp` or `val2` variable name.

So I moved `new_counter` after the loop, and do the `+` twice :shrug: 

I still believe `uninitialized` was fine. It was merely out constraining the variable scope and it was never read before it was defined. In fact, a linter could prove it.

**Comment by ysbaddaden** [comment]
AFAIK it should now be impossible to fail to wake the waiting fibers or leave the waitgroup in a confusing state: the counter saturates to a _negative_ number and can't return to a positive number anymore; waiting fibers are always resumed (once) when the counter reaches zero or below.

I can't think of any scenario where we'd end up with a deadlock.

I can still think of race conditions, though:

1. counter reaches zero
2. waiting fibers are enqueued
3. counter reaches a negative number

Depending on when the fibers are resumed, some may return successfully (zero counter) while some may raise (negative counter), yet at least one fiber will raise (the one decrementing the counter below zero), so the error shouldn't go unnoticed.

**Review by RX14 — APPROVED** [review]
This looks great to me, and I agree with you about the remaining race condition.

**Comment by jkthorne** [comment]
I cannot wait to use this over some Channels that I have.

**Comment by bararchy** [comment]
Same! very excited for this one :tada:

File: src/wait_group.cr
Comment (straight-shoota) [reply src/wait_group.cr]: Actually, there should be no need for `not_nil!`. The compiler is capable of figuring out that the loop runs at least once and it assigns to `new_counter`, so it's typed without `Nil` afterwards.
So literally just changing `uninitialized Int32` to `nil` should've worked. I'm fine with the refactored version, though.

**Review by straight-shoota — APPROVED** [review]

File: src/wait_group.cr
Comment (bararchy) [reply src/wait_group.cr:94]: @ysbaddaden now that it's in and merged, are you planning to make the followup PR? :eye:

**Pull request merged**
